common/src/chain/config/emission_schedule.rs

```diff
@@ -0,0 +1,336 @@
+
+    /// Construct from the number atomic units
+    pub const fn from_atoms(n: u128) -> Self {
+        Self(Amount::from_atoms(n))
+    }
+
+    /// Construct from the number of MLTs
+    pub const fn from_mlt(n: u64) -> Self {
+        // Since the argument is u64 and number of atoms in 1 MLT is <= u64::MAX,
+        // the result is guaranteed to fit into the internal representation of Amount (u128)
+        static_assertions::const_assert!(Mlt::ATOMS_PER_MLT <= u64::MAX as u128);
+        Self(Amount::from_atoms(n as u128 * Mlt::ATOMS_PER_MLT))
+    }
+
+    /// Convert the number of atoms to Amount
+    pub const fn to_amount_atoms(self) -> Amount {
+        self.0
+    }
+
+    /// Multiplication, saturated at [Mlt::MAX]
+    pub fn saturating_mul(self, rhs: u128) -> Self {
+        Self::from_atoms(self.0.into_atoms().saturating_mul(rhs))
+    }
+
+    /// Addition, saturated at [Mlt::MAX]
+    pub fn saturating_add(self, rhs: Self) -> Self {
+        Self::from_atoms(self.0.into_atoms().saturating_add(rhs.0.into_atoms()))
+    }
+}
+
+impl std::ops::Add for Mlt {
+    type Output = Option<Self>;
+    fn add(self, rhs: Self) -> Option<Self> {
+        (self.0 + rhs.0).map(Self)
+    }
+}
+
+impl std::ops::Sub for Mlt {
+    type Output = Option<Self>;
+    fn sub(self, rhs: Self) -> Option<Self> {
+        (self.0 - rhs.0).map(Self)
+    }
+}
+
+impl std::ops::Mul<u128> for Mlt {
+    type Output = Option<Self>;
+    fn mul(self, rhs: u128) -> Option<Self> {
+        (self.0 * rhs).map(Self)
+    }
+}
+
+/// Internal emission schedule representation
+pub type EmissionScheduleFn = dyn Fn(BlockHeight) -> Mlt + Sync + Send + 'static;
+
+/// Emission schedule, characterized by function from block height to total supply at that point.
+///
+/// The function has to be a monotonic non-decreasing function. The function also has to
+/// eventually flatten out to a constant function if [Self::final_supply] is required to make any
+/// sense. The point at which the function becomes constant is where the block rewards run out.
+#[derive(Clone)]
+pub struct EmissionSchedule(std::sync::Arc<EmissionScheduleFn>);
+
+impl EmissionSchedule {
+    /// Construct an emission schedule from a function.
+    ///
+    /// Be careful to maintain invariants as specified in [EmissionSchedule] docs
+    pub fn from_fn(f: impl 'static + Fn(BlockHeight) -> Mlt + Sync + Send) -> Self {
+        Self(std::sync::Arc::new(f))
+    }
+
+    /// Construct an emission schedule from a function. See also [Self::from_fn]
+    pub fn from_dyn_fn(f: Box<EmissionScheduleFn>) -> Self {
+        Self(f.into())
+    }
+
+    /// Construct an emission schedule from yearly block award table
+    ///
+    /// This constructor takes:
+    ///  * The initial supply (premine amount)
+    ///  * Yearly per-block rewards, as a vector.
+    ///    In years outside of the range of the vector, the reward is assumed to be zero.
+    ///  * The target block interval
```

Comment:
Should there be a mention of the precondition that an interval cannot contain sub-seconds?

Reply:
Good point. Added.